File: packages/shared/schemas/lead.ts
import { z } from "zod";

export const CreateLeadSchema = z.object({
  name: z.string().min(1, "Required"),
  email: z.string().min(1, "Required"),
  phone: z.string().min(1, "Required"),
  company: z.string().min(1, "Required"),
  project_type: z.string().min(1, "Required"),
  budget_range: z.string().min(1, "Required"),
  message: z.string().optional(),
  source: z.string().min(1, "Required"),
  status: z.string().min(1, "Required"),
  internal_notes: z.string().optional(),
});

export const UpdateLeadSchema = z.object({
  name: z.string().min(1, "Required").optional(),
  email: z.string().min(1, "Required").optional(),
  phone: z.string().min(1, "Required").optional(),
  company: z.string().min(1, "Required").optional(),
  project_type: z.string().min(1, "Required").optional(),
  budget_range: z.string().min(1, "Required").optional(),
  message: z.string().optional(),
  source: z.string().min(1, "Required").optional(),
  status: z.string().min(1, "Required").optional(),
  internal_notes: z.string().optional(),
});

export type CreateLeadInput = z.infer<typeof CreateLeadSchema>;
export type UpdateLeadInput = z.infer<typeof UpdateLeadSchema>;

File: apps/api/internal/models/lead.go
package models

import (
	"time"

	"gorm.io/datatypes"
	"gorm.io/gorm"

	"megagig-software-solution/apps/api/internal/ids"
)

// Lead represents a lead in the system.
type Lead struct {
	ID                 string                      `gorm:"primarykey;size:36" json:"id"`
	Name               string                      `gorm:"size:255" json:"name" binding:"required"`
	Email              string                      `gorm:"size:255" json:"email" binding:"required"`
	Phone              string                      `gorm:"size:255" json:"phone" binding:"required"`
	Company            string                      `gorm:"size:255" json:"company" binding:"required"`
	ProjectType        string                      `gorm:"size:255" json:"project_type" binding:"required"`
	BudgetRange        string                      `gorm:"size:255" json:"budget_range" binding:"required"`
	ServicesInterested datatypes.JSONSlice[string] `gorm:"type:json" json:"services_interested"`
	Message            string                      `gorm:"type:text" json:"message"`
	Source             string                      `gorm:"size:255" json:"source" binding:"required"`
	Status             string                      `gorm:"size:255" json:"status" binding:"required"`
	InternalNotes      string                      `gorm:"type:text" json:"internal_notes"`
	Version            int                         `gorm:"not null;default:1" json:"version"`
	CreatedAt          time.Time                   `json:"created_at"`
	UpdatedAt          time.Time                   `json:"updated_at"`
	DeletedAt          gorm.DeletedAt              `gorm:"index" json:"-"`
}

// BeforeCreate generates a UUID before inserting.
func (m *Lead) BeforeCreate(tx *gorm.DB) error {
	if m.ID == "" {
		m.ID = ids.New()
	}
	return nil
}

// BeforeUpdate increments Version so offline clients can detect server-side updates.
func (m *Lead) BeforeUpdate(tx *gorm.DB) error {
	tx.Statement.SetColumn("version", gorm.Expr("version + 1"))
	return nil
}
